flink-python/src/main/java/org/apache/flink/python/ProcessEnvironmentManager.java

```diff
@@ -0,0 +1,353 @@
+	 * @throws IllegalArgumentException
+	 */
+	private void prepareWorkingDir() throws IOException, IllegalArgumentException {
+
+		// Extract internal python libs and udf runner script.
+		new File(basicDirectory).mkdirs();
+		ResourceUtil.extractBasicDependenciesFromResource(
+			basicDirectory,
+			this.getClass().getClassLoader(),
+			"",
+			false);
+
+		for (Map.Entry<String, String> entry : dependencyManager.getFilesInPythonPath().entrySet()) {
+			// The origin file name will be wiped during the transmission in Flink Distributed Cache
+			// and replaced by a unreadable character sequence, now restore their origin name.
+			// Every python file will be placed at :
+			// ${basicDirectory}/python_files_dir/${distributedCacheFileName}/${originFileName}
+			String distributedCacheFileName = new File(entry.getKey()).getName();
+			String actualFileName = entry.getValue();
+			Path target = FileSystems.getDefault().getPath(filesDirectory,
+				distributedCacheFileName, actualFileName);
+			if (!target.getParent().toFile().mkdirs()) {
+				throw new IllegalArgumentException(
+					String.format("Can not create the directory: %s !", target.getParent().toString()));
+			}
+			Path src = FileSystems.getDefault().getPath(entry.getKey());
+			try {
+				Files.createSymbolicLink(target, src);
+			} catch (IOException e) {
+				LOG.warn(String.format("Can not create the symbolic link of: %s, "
+					+ "the link path is %s, fallback to copy.", src, target), e);
+				FileUtils.copy(new org.apache.flink.core.fs.Path(src.toUri()),
+					new org.apache.flink.core.fs.Path(target.toUri()), false);
+			}
+		}
+
+		for (Map.Entry<String, String> entry : dependencyManager.getArchives().entrySet()) {
+			UnzipUtil.extractZipFileWithPermissions(
+				entry.getKey(), joinPath(archivesDirectory, entry.getValue()));
+		}
+
+		File requirementsDirectoryFile = new File(requirementsDirectory);
+		if (!requirementsDirectoryFile.exists()) {
+			if (!requirementsDirectoryFile.mkdirs()) {
+				throw new IllegalArgumentException(
+					String.format("Creating the requirements target directory: %s failed!",
+						requirementsDirectory));
+			}
+		} else {
+			if (!requirementsDirectoryFile.isDirectory()) {
+				throw new IllegalArgumentException(
+					String.format("The requirements target directory path: %s is not a directory!",
+						requirementsDirectory));
+			}
+		}
+	}
+
+	private Thread registerShutdownHook() {
+		Thread thread = new Thread(() -> FileUtils.deleteDirectoryQuietly(new File(basicDirectory)));
+		Runtime.getRuntime().addShutdownHook(thread);
+		return thread;
+	}
+
+	@VisibleForTesting
+	String getBasicDirectory() {
+		return basicDirectory;
+	}
+
+	static String joinPath(String... args) {
+		return String.join(File.separator, args);
+	}
+
+	private static void appendToPythonPath(Map<String, String> env, String libs) {
+		String systemPythonPath = env.get("PYTHONPATH");
+		if (systemPythonPath != null && !systemPythonPath.isEmpty()) {
```

Comment:
What happens if systemPythonPath is empty?

Reply:
If we not check empty there will be an extra path separator at the end of PYTHONPATH.